src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/SQLEngineAvroToStructuredTransformer.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.gcp.bigquery.sqlengine;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.etl.api.engine.sql.SQLEngineException;
+import io.cdap.plugin.gcp.bigquery.source.BigQueryAvroToStructuredTransformer;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import javax.annotation.Nullable;
+
+/**
+ * Create StructuredRecords from GenericRecords when pulling records from BigQuery.
+ *
+ * The following transformations are supported:
+ * BigQuery INT64 -> Integer (with Overflow check)
+ * BigQuery FLOAT64 -> Float (with Overflow check)
+ *
+ * Otherwise, field conversion gets delegated to the {@link BigQueryAvroToStructuredTransformer}
+ */
+public class SQLEngineAvroToStructuredTransformer extends BigQueryAvroToStructuredTransformer
+  implements Serializable {
+
+  private static final String LONG_TO_INTEGER_CONVERSION_ERROR =
+    "Could not convert Long value '%d' into Integer when reading from BigQuery";
+  private static final String INTEGER_CONVERSION_ERROR =
+    "Could not convert Object '%s' into Integer when reading from BigQuery";
+  private static final String DOUBLE_TO_FLOAT_CONVERSION_ERROR =
+    "Could not convert Double '%f' into Float when reading from BigQuery";
+  private static final String FLOAT_CONVERSION_ERROR =
+    "Could not convert Object '%s' into Float when reading from BigQuery";
+
+  @Override
+  @Nullable
+  protected Object convertField(Object field, Schema fieldSchema) throws IOException {
+    if (field == null) {
+      return null;
+    }
+
+    // Handle Int types
+    if (fieldSchema.getType() == Schema.Type.INT) {
+      return mapInteger(field);
+    }
+
+    // Handle float types
+    if (fieldSchema.getType() == Schema.Type.FLOAT) {
+      return mapFloat(field);
+    }
+
+    // Handle Strings that are stored as a Byte Buffer.
+    if (fieldSchema.getType() == Schema.Type.STRING && field instanceof ByteBuffer) {
+      return StandardCharsets.UTF_8.decode((ByteBuffer) field).toString();
+    }
+
+    return super.convertField(field, fieldSchema);
+  }
+
+  @VisibleForTesting
+  protected Integer mapInteger(Object field) throws SQLEngineException {
```

Comment:
will mapInteger ever receive an Integer as input?

Reply:
No, the reason I had to implement this function is because BigQuery stores integers as INT64 (which get mapped to Long when reading form the Avro input).

Same for Floats, as they are stored in Float64 (which get read as Double from Avro)